firebase-firestore/src/main/java/com/google/firebase/firestore/local/RemoteDocumentCache.java

```diff
@@ -40,8 +41,9 @@
    * for the key, it will be replaced.
    *
    * @param maybeDocument A Document or NoDocument to put in the cache.
+   * @param readTime The time at which the document was read or committed.
    */
-  void add(MaybeDocument maybeDocument);
+  void add(MaybeDocument maybeDocument, SnapshotVersion readTime);
```

Comment:
Dumb question: why not make `readTime` a part of the `MaybeDocument`?

Reply:
That was the road I went down on at first - it generates a significant amount of changes to plumb through a value that is only ever read directly at source.

Reply:
I'm sort of surprised there are that many changes. Did you happen to commit it so I can see?

For example: all our tests could use the same snapshot for both versions by default and only specify a second timestamp when the difference really matters. That means nearly all tests would be unaffected.

Oh, maybe the issue is that the read_time comes from the TargetChange and is not intrinsically a part of the document?

Reply:
There are at least two additional challenges:
- If I add it to MaybeDocument, I have to change it for Document, UnknownDocument and NoDocument. At least for NoDocuments it is throwaway work as these are never written to the cache.
- MaybeDoocument currently uses the generic term `version`. This works well with all subtypes, but it doesn't quite work if introduce another kind of version.

I never actually got as far as to see how we would construct the individual documents. By exposing the read time only as an input to the RemoteDocumentCache there are only two callsites that need changing, which won me over.